src/pkg/user/service/service.go

```diff
@@ -22,77 +28,247 @@ type Service struct {
+	}
+
+	procRequest.AddEthereumTx(proc.TxDeleteDoc, txHash)
+
+	return nil
+}
+
+func (s *Service) Login(ctx context.Context, user *model.UserAuthRequest) (err error) {
+	ehrSystemID := ctx.(*gin.Context).GetString("ehrSystemID")
+	address, pwdHash, err := s.getUserAddressAndHash(ehrSystemID, user.UserID, user.Password)
+
+	if err != nil {
+		return err
+	}
+
+	userHash, err := s.Infra.Index.GetUserPasswordHash(ctx, address)
+	if err != nil {
+		return fmt.Errorf("Login.GetUserPasswordHash error: %w", err)
+	}
+
+	if !bytes.Equal(pwdHash, userHash) {
+		return errors.ErrFieldIsIncorrect("Password")
+	}
+
+	return nil
+}
+
+func (s *Service) getUserAddressAndHash(ehrSystemID, userID, password string) (eth_common.Address, []byte, error) {
+	_, userPrivateKey, err := s.Infra.Keystore.Get(userID)
+	if err != nil {
+		return eth_common.Address{}, nil, fmt.Errorf("Keystore.Get error: %w userID %s", err, userID)
 	}
 
 	privateUserKey := userPrivateKey[:]
 
 	privateKey, err := crypto.ToECDSA(privateUserKey)
 	if err != nil {
-		return fmt.Errorf("crypto.ToECDSA error: %w userID %s", err, user.UserID)
+		return eth_common.Address{}, nil, fmt.Errorf("crypto.ToECDSA error: %w userID %s", err, userID)
 	}
 
 	address := crypto.PubkeyToAddress(privateKey.PublicKey)
-	ehrSystemID := ctx.(*gin.Context).GetString("ehrSystemID")
 
-	pwdHash, err := s.generateHash(user.UserID, ehrSystemID, user.Password)
+	pwdHash, err := s.generateHash(privateUserKey, userID, ehrSystemID, password)
 	if err != nil {
-		return fmt.Errorf("register s.generateHash error: %w userID %s, password: %s", err, user.UserID, user.Password)
+		return eth_common.Address{}, nil, fmt.Errorf("register s.generateHash error: %w userID %s, password: %s", err, userID, password)
 	}
+	return address, pwdHash, nil
+}
 
-	requestID := ctx.(*gin.Context).GetString("reqId")
+// method should be idempotent
+func (s *Service) generateHash(salt []byte, phrases ...string) ([]byte, error) {
+	hash := strings.Join(phrases, "")
 
-	txHash, err := s.Infra.Index.UserAdd(requestID, address, user.UserID, user.Role, pwdHash)
+	pwdHash, err := scrypt.Key([]byte(hash), salt, N, r, p, keyLen)
 	if err != nil {
-		return fmt.Errorf("Index.DeleteDoc error: %w", err)
+		return nil, fmt.Errorf("scrypt.Key error: %w", err)
 	}
 
-	procRequest.AddEthereumTx(proc.TxDeleteDoc, txHash)
+	return pwdHash, nil
+}
+
+func (s *Service) CreateToken(userID string) (*TokenDetails, error) {
+	td := &TokenDetails{}
+	td.AtExpires = time.Now().Add(common.JWTExpires).Unix()
+	td.AccessUUID = uuid.New().String()
+
+	td.RtExpires = time.Now().Add(common.JWTRefreshExpires).Unix()
+	td.RefreshUUID = td.AccessUUID + "++" + userID
+
+	var err error
+	//Creating Access Token
+	accessTokenSecret, _, err := s.Infra.Keystore.Get(userID)
+	if err != nil {
+		return nil, fmt.Errorf("CreateToken Keystore.Get error: %w userID %s", err, userID)
+	}
 
+	refreshTokenSecret, _, err := s.Infra.Keystore.Get(userID + "_refresh")
```

Comment:
1. Keystore stores ECDSA keys of users according to userID. You do not need to add "_refresh" to them
2. Using publicKey as a tokenSecret is a very bad idea

Reply:
This must be removed. Use ecdsaAccessKey instead of ecdsaRefreshKey.
And rename ecdsaAccessKey to userECDSAKey